[CI] Build example applications, wrapper and vboxwrapper for OSX.

AenBleidd opened the pull request:
This fixes #4892.

Signed-off-by: Vitalii Koshura <lestat.de.lionkur@gmail.com>

Fixes #

**Description of the Change**
<!-- We must be able to understand the design of your change from this description. -->

**Alternate Designs**
<!-- Explain what other alternates were considered and why the proposed version was selected -->

**Release Notes**
<!--
Please describe the changes in a single line that explains this improvement in terms that a user can understand.
This text will be used in BOINC's release notes.
If this change is not user-facing or notable enough to be included in release notes you may use the string "N/A" here. -->

CharlieFenton commented:
@AenBleidd I haven't looked at the other samples, but _example_app/Makefile_mac_ is obviously out of date. Fro example, it has: `FREETYPE_DIR = $(BOINC_DIR)/../freetype-2.4.10` when the current dependent version specified in _mac_build/dependencyNames.sh_ is freetype-2.11.0. I suspect this may be the reason for the build failure.

CharlieFenton commented:
@AenBleidd Also, in example_app/MakeMacExample.sh you may want to update `-DMAC_OS_X_VERSION_MAX_ALLOWED=1070` to `-DMAC_OS_X_VERSION_MAX_ALLOWED=101300`, and change `export MACOSX_DEPLOYMENT_TARGET=10.7` to `export MACOSX_DEPLOYMENT_TARGET=10.13` to eliminate the warnings "was built for newer macOS version (10.13) than being linked (10.7)". 

You may want to make similar changes for the other two samples. But since these are only warnings, all 3 builds should succeed without these changes.

AenBleidd commented:
@CharlieFenton, thanks for the hint. I'll slowly step by step fix build issues and update these scripts. When done - I'll definitely ask to review them. Thank you in advance.

CharlieFenton commented:
> Also, in _example_app/MakeMacExample.sh_ you may want to update `-DMAC_OS_X_VERSION_MAX_ALLOWED=1070` to `-DMAC_OS_X_VERSION_MAX_ALLOWED=101300`, and change `export MACOSX_DEPLOYMENT_TARGET=10.7` to `export MACOSX_DEPLOYMENT_TARGET=10.13` to eliminate the warnings "was built for newer macOS version (10.13) than being linked (10.7)".

I forgot to mention: if you do this, you should also change `-DMAC_OS_X_VERSION_MIN_REQUIRED=1070` to `-DMAC_OS_X_VERSION_MIN_REQUIRED=101300`. And similar changes for the other two samples.

AenBleidd commented:
Hello @CharlieFenton, it would be nice if you could review this and test (if possible).
This is a minor change, and doesn't interfere with BOINC release process, so it's not urgent.
Thank you in advance

CharlieFenton commented:
@AenBleidd These changes are more extensive than i expected, so it will take me a little time to review, but I will try to get it done within the next week.

CharlieFenton commented:
**in _BuildMacBOINC.sh_**:

While _BuildMacBOINC-CI.sh_ calls _BuildMacBOINC-CI.sh_ to run CI builds,  _BuildMacBOINC-CI.sh_ has a different purpose from _BuildMacBOINC-CI.sh_. It is intended for:
* People who want to build the client and BOINC Manager themselves but don't want to learn how to use the Xcode IDE. It allows them to build from the command line.
* Projects that want to build the BOINC libraries from the command line.
Neither of these need the Zip samples, so these should be built **only** if the argument `-zipapps` is passed. The zip samples should not be built for `-all` unless `-zipapps` is also passed. 

In your modified _BuildMacBOINC-CI.sh_, _libboinc_zip.a_ is no longer built for the default case (when there are no arguments.) This was the purpose of the logic involving the variable `buildzip` which you removed. Please fix this.

Also for the reasons above:
* the example_app samples should not be built for `-all` unless `-uc2` is also passed to _BuildMacBOINC.sh_.
* vboxwrapper should not be built for -all unless `-vboxwrapper` is also passed to _BuildMacBOINC.sh_.

**in buildMacMakefiles-CI.sh**:

Is it necessary to build the libraries again via `BuildMacBOINC.sh -lib` if they were already built from _BuildMacBOINC-CI.sh_? 

The Xcode project for target _jpeg_ already know where to find the source and header files, so `-setting HEADER_SEARCH_PATHS "../samples/jpeglib"` should not be necessary on line 73.

**In MakeMacExample.sh**:

Since you changed the "exit" statements to "return", I get this warning / error when I invoke it with `sh [path]/MakeMacExample.sh`:
`line 193: return: can only `return' from a function or sourced script`

So please change line 46 to say `source [path]/MakeMacExample.sh` rather than `sh [path]/MakeMacExample.sh`

But if I invoke it with `source [path]/MakeMacExample.sh`, I get:
```
ld: warning: directory not found for option '-L/Volumes/Dev/BOINC_GIT/boinc_vko_add_more_osx_ci/samples/example_app/3rdParty/buildCache/mac/lib'
ld: library not found for -lboinc_zip
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
Because you are always setting `PREFIX` either to `$cache_dir` or to `$(pwd)/3rdParty/buildCache/mac`, this cannot be built outside of CI. The default needs to be for `PREFIX` to be an empty string.

**Also in example_app**:

Your code does not test building with _Makefile_mac2_; please see _example_app/ReadMe.txt_. Note that, like Makefile_mac, this also needs to change _freetype-2.4.10_ to _freetype-2.11.0_. Also note that it does not currently use `$(PREFIX)`.

**samples/mac_build/UpperCase2.xcodeproj**:

You changed the header search paths and other _ldflags_ in such a way that it no longer works on my Mac. These needed to just change "freetype-2.4.10" to "freetype-2.11.0", but I don't know if that will work for the CI builds. The invocation from  _BuildMacBOINC-CI.sh_ might need to make a separate call to _BuildMacBOINC-CI.sh_ to build just this sample with custom values for `-setting USER_HEADER_SEARCH_PATHS` and `-setting LIBRARY_SEARCH_PATHS`.

It should also be changed to build arm64 as well as x86_64 executables. I'll email you privately a fixed version that works on my Mac.

**in BuildMacVboxWrapper.sh**:

You changed some exit statements to return but not all of them. Please fix.

Please change line 44 to say `source [path]/BuildMacVboxWrapper` rather than `sh [path]/BuildMacVboxWrapper`

You have not added `$(PREFIX)` to _vboxwrapper/makefile_mac_

Because you are always setting `PREFIX` either to `$cache_dir` or to `$(pwd)/3rdParty/buildCache/mac`, this cannot be built outside of CI once you add `$(PREFIX)` to _vboxwrapper/makefile_mac_.

**in vboxwrapper.xcodeproj**:

It needs to build arm64 as well as X86_64. I'll email you a fixed version that works on my Mac.

**In BuildMacWrapper.sh**:

On line 28, you wrote "Updated 1/17/23 to sign binaries". Where does it do that, and where does it get the signing credentials?

Please change line 41 to say `source [path]/BuildMacWrapper` rather than `sh [path]/BuildMacWrapper`.

You have not added `$(PREFIX)` to _wrapper/makefile_mac_

Because you are always setting `PREFIX` either to `$cache_dir` or to `$(pwd)/3rdParty/buildCache/mac`, this cannot be built outside of CI once you add `$(PREFIX)` to _wrapper/makefile_mac_.

CharlieFenton reviewed (changes requested):
Please see my detailed comments [here](https://github.com/BOINC/boinc/pull/5056#issuecomment-1403642492)

AenBleidd commented:
> In your modified BuildMacBOINC-CI.sh, libboinc_zip.a is no longer built for the default case (when there are no arguments.) This was the purpose of the logic involving the variable buildzip which you removed. Please fix this.

Not sure you're correct here.

Yes, for the default case `libboinc_zip.a` should be built. And according to the logic in the files, it is building.
Old logic:
`buildzip` is `true` when building either 
- when `default (no params)` then is later interpreted as `-all`:
https://github.com/BOINC/boinc/blob/ba026ec81922b6982403b9f3b077e07e6089e98d/mac_build/BuildMacBOINC.sh#L127
- when `-all` or `-lib`: https://github.com/BOINC/boinc/blob/ba026ec81922b6982403b9f3b077e07e6089e98d/mac_build/BuildMacBOINC.sh#L182

I changed to code a little bit but it still builds `libboinc_zip.a` for these 3 cases including a new one when building `zipapps` (please be aware that `default (no params)` is still exposed to `-all`:
```bash
    if [ "${buildall}" = "1" ] || [ "${buildlibs}" = "1" ] || [ "${buildzipapps}" = "1" ]; then
        eval "xcodebuild -project ../zip/boinc_zip.xcodeproj -target boinc_zip -configuration ${style} -sdk \"${SDKPATH}\" ${doclean} build  ${uselibcplusplus} ${theSettings}"
        result=$?
    fi
```

AenBleidd commented:
> Neither of these need the Zip samples, so these should be built only if the argument `-zipapps` is passed. The zip samples should not be built for `-all` unless `-zipapps` is also passed.

Fixed.

> the example_app samples should not be built for `-all` unless `-uc2` is also passed to BuildMacBOINC.sh.

Fixed.

> Is it necessary to build the libraries again via `BuildMacBOINC.sh -lib` if they were already built from BuildMacBOINC-CI.sh?

Yes, since this script `buildMacMakefiles-CI.sh` is an alternative way of building custom targets (UC2, wrapper and vboxwrapper). This script is mostly for running on CI only (ant just testing different stuff), so in this flow it will build these libraries the first time, and this is ok even if this is a duplicate.
To summarize, this is by intention.

AenBleidd commented:
> So please change line 46 to say `source [path]/MakeMacExample.sh` rather than `sh [path]/MakeMacExample.sh`

Fixed

> The default needs to be for `PREFIX` to be an empty string.

Fixed

AenBleidd commented:
> You changed some exit statements to return but not all of them. Please fix.

Fixed.

> Please change line 44 to say `source [path]/BuildMacVboxWrapper` rather than `sh [path]/BuildMacVboxWrapper`

Fixed.

AenBleidd commented:
> You have not added `$(PREFIX)` to vboxwrapper/makefile_mac

Fixed.

> this cannot be built outside of CI once you add `$(PREFIX)` to vboxwrapper/makefile_mac.

Fixed.

> Please change line 41 to say `source [path]/BuildMacWrapper` rather than `sh [path]/BuildMacWrapper`.

Done

> You have not added `$(PREFIX)` to wrapper/makefile_mac

Done.

> this cannot be built outside of CI once you add `$(PREFIX)` to wrapper/makefile_mac.

Done.

AenBleidd commented:
> On line 28, you wrote "Updated 1/17/23 to sign binaries". Where does it do that, and where does it get the signing credentials?

I just used the piece of code from the `BuildMacVboxWrapper.sh` that uses `"${HOME}/BOINCCodeSignIdentities.txt"` file.
Actually, on CI this doesn't work since this file is not present there, so this step is just skipped, but you can test this locally, I believe. I think it's important to sign `wrapper` as well as `vboxwrapper` (at least for windows we publish both Windows binaries signed).

CharlieFenton commented:
> I just used the piece of code from the BuildMacVboxWrapper.sh that uses "${HOME}/BOINCCodeSignIdentities.txt" file.
Actually, on CI this doesn't work since this file is not present there, so this step is just skipped, but you can test this locally, I believe. I think it's important to sign wrapper as well as vboxwrapper (at least for windows we publish both Windows binaries signed).

OK I will look at this next week. I have no moe time this week.

AenBleidd commented:
> OK I will look at this next week. I have no more time this week.

I'm not done with this PR yet.
I left this comments mostly to not forget what I already fixed and what is left :-D
Sorry for spamming.
Once done, I'll make this PR `Ready for review` again and let you know.
Thank you for your time. @CharlieFenton, I really appreciate this!

AenBleidd commented:
@CharlieFenton, PR is ready for your review, please check it.
`buildjpeg.sh` was removed by intention because it's obsolete and its functionality was moved to XCode project file.

CharlieFenton commented:
Thank you. I will review some time this week.

CharlieFenton commented:
@AenBleidd Thank you for the changes. We are almost there! A few more items:

**In BuildMacBOINC.sh**:

The usage message (in response to the `-help` argument or a bad argument is missing the following options:
`[-libstdc++]`
`[-zipapps]`
`[-uc2]`
`[-vboxwrapper]`

You have this comment for the `[-zipapps]`, `[-uc2]` and `[-vboxwrapper]` arguments:
> If run as a single target and not in sequence when libs are already build it will fail to build without -lib parameter specified

It might be better to automatically set `$buildlibs=1` if any of these 3 arguments are passed. What do you think?

**In example_app/ReadMe.txt**:

Please change line 16 from `sh MakeMacExample.sh` to `source MakeMacExample.sh`

AenBleidd commented:
> It might be better to automatically set `$buildlibs=1` if any of these 3 arguments are passed. What do you think?

Basically, this is not good for ci, because it will build libs 3 more times, but from the user pov - this is definitely better, so I fixed that.

Both other comments are fixed as well.

@CharlieFenton, please re-review.

CharlieFenton commented:
@AenBleidd I have approved this PR, but I see one of the CI builds is failing, so I have not merged it.

AenBleidd commented:
> @AenBleidd I have approved this PR, but I see one of the CI builds is failing, so I have not merged it.

Yeah, that was a CI issue during downloading of some tool. I rerun it and everything is fine, so I'm gonna merge it now.
@CharlieFenton, thank you for your review, your time and your assistance.
I really appreciate this.
